modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksStorageUtils.java

```diff
@@ -129,47 +131,57 @@ static long revisionFromRocksKey(byte[] rocksKey) {
         return (long) LONG_ARRAY_HANDLE.get(rocksKey, 0);
     }
 
-    /**
-     * Builds a value from a byte array.
-     *
-     * @param valueBytes Value byte array.
-     * @return Value.
-     */
+    /** Converts from a byte array to a {@link Value}. */
     static Value bytesToValue(byte[] valueBytes) {
-        // At least an 8-byte update counter and a 1-byte boolean
-        assert valueBytes.length > Long.BYTES;
+        // At least an 8-byte update counter, 8-bytes operation timestamp and a 1-byte boolean.
+        assert valueBytes.length > 2 * Long.BYTES;
+
+        var pos = 0;
 
         // Read an update counter (8-byte long) from the entry.
-        long updateCounter = (long) LONG_ARRAY_HANDLE.get(valueBytes, 0);
+        long updateCounter = (long) LONG_ARRAY_HANDLE.get(valueBytes, pos);
+        pos += Long.BYTES;
+
+        // Read an operation timestamp (8-byte long) from the entry.
+        long operationTimestamp = (long) LONG_ARRAY_HANDLE.get(valueBytes, pos);
+        pos += Long.BYTES;
 
         // Read a has-value flag (1 byte) from the entry.
-        boolean hasValue = valueBytes[Long.BYTES] != 0;
+        boolean hasValue = valueBytes[pos] != 0;
+        pos += Byte.BYTES;
 
         byte[] val;
         if (hasValue) { // Copy the value.
-            val = Arrays.copyOfRange(valueBytes, Long.BYTES + 1, valueBytes.length);
+            val = Arrays.copyOfRange(valueBytes, pos, valueBytes.length);
         } else { // There is no value, mark it as a tombstone.
             val = TOMBSTONE;
         }
 
-        return new Value(val, updateCounter);
+        return new Value(val, updateCounter, hybridTimestamp(operationTimestamp));
     }
 
     /**
-     * Adds an update counter and a tombstone flag to a value.
+     * Converts the contents of a {@link Value} to a byte array.
      *
-     * @param value         Value byte array.
+     * @param value Value byte array.
      * @param updateCounter Update counter.
+     * @param operationTimestamp Operation timestamp.
      * @return Value with an update counter and a tombstone.
      */
-    static byte[] valueToBytes(byte[] value, long updateCounter) {
-        var bytes = new byte[Long.BYTES + Byte.BYTES + value.length];
+    static byte[] valueToBytes(byte[] value, long updateCounter, HybridTimestamp operationTimestamp) {
```

Comment:
Why do we need to store timestamp in a value? Imagine that there is a putAll() with 100 entries; all of them will get the same revision and timestamp, and we will store the timestamp 100 times instead of just one. We don't save the revision once per value, and we should not do this for timestamps, as it seems. When reading, we know the revision, and we can obtain the timestamp by that revision.

Reply:
In [IGNITE-23307](https://issues.apache.org/jira/browse/IGNITE-23307), after discussion with colleagues, we decided that methods for working with revisions and timestamps should throw `CompactedException`. But since after compacting metastorage we will be able to receive keys even with a compacted revision (for example, only one version by key from the beginning of time), then for other components, as we decided, it will be useful to also return the timestamp. Therefore, we will store the timestamp in the value; I think storing it in the key is redundant.

Reply:
Oh, so the mapping from revisions to timestamps will also be compacted, I see. Could you please mention this in the Jira issue?